librosa/core/pitch.py

```diff
@@ -335,3 +338,281 @@ def piptrack(y=None, sr=22050, S=None, n_fft=2048, hop_length=None,
                                   + dskew[idx[:, 0], idx[:, 1]])
 
     return pitches, mags
+
+
+def yin(y, sr=22050, win_length=1024, hop_length=None, fmin=55, fmax=900, peak_threshold=0.1):
+    '''Fundamental frequency (F0) estimation. [1]_
+
+    .. [1] De Cheveigné, A., & Kawahara, H. (2002).
+        "YIN, a fundamental frequency estimator for speech and music."
+        The Journal of the Acoustical Society of America, 111(4), 1917-1930.
+
+     Parameters
+     ----------
+     y : np.ndarray [shape=(n,)]
+         audio time series.
+
+     sr : number > 0 [scalar]
+         sampling rate of `y` in Hertz.
+
+     win_length : int [scalar]
+         length of the window in samples.
+         By default, win_length=1024 corresponds to a time scale of 46 ms at
+         a sampling rate of 22050 Hz.
+
+     hop_length : None or int > 0 [scalar]
+         number of audio samples between adjacent pYIN predictions.
+         If `None`, defaults to `frame_length // 4`.
+
+     fmin: number > 0 [scalar]
+         minimum frequency in Hertz.
+
+     fmax: number > 0 [scalar]
+         maximum frequency in Hertz.
+
+     peak_threshold: number > 0 [scalar]
+         absolute threshold for peak estimation.
+
+     Returns
+     -------
+     f0: np.ndarray [shape=(n_frames,)]
+         time series of fundamental frequencies in Hertz.
+
+     Examples
+     --------
+     Computing a fundamental frequency curve from an audio input:
+     >>> y = librosa.chirp(440, 880, duration=7.0)
+     >>> yin(y)
+     array([441.15076669, 441.6640896 , 442.16219657, ...,
+        871.59470743, 872.63699131, 873.66696896])
+     '''
+    # Set the default hop if it is not already specified.
+    if hop_length is None:
+        hop_length = int(win_length // 4)
+
+    # Check that audio is valid.
+    util.valid_audio(y, mono=True)
+
+    # Autocorrelation.
+    y_frames = util.frame(y, frame_length=2*win_length, hop_length=hop_length)
```

Comment:
For consistency with the rest of the library, there should be a frame-centering option here, which controls padding prior to framing.  Otherwise, the resulting time grid will not be consistent with default spectrogram parameters.

Reply:
I added the frame-centering option. It works fine for `yin` but for `pyin` it introduces an annoying effect of lowering the trough probabilities in the first frame causing `pyin` to classify it as unvoiced. For most practical applications that probably won't be an issue but I had to turn centering off for the tests. I hope that's okay?

Reply:
>  it introduces an annoying effect of lowering the trough probabilities in the first frame causing `pyin` to classify it as unvoiced. For most practical applications that probably won't be an issue but I had to turn centering off for the tests. I hope that's okay?

Yeah, I think that's probably fine.  I wonder if there's an elegant hack around it though.

The windowed autocorrelation algorithm assumes left-aligned frames, and takes the first `n/2+1` as the template.  This is a little weird when we have centered frames, as on the first one, those first samples are reflect-pads (or zero-pads?) of the first frame.  Maybe it would make more sense to use the center `n/2+1` samples for the template if we're in centered mode?

Reply:
I'll give it a try.

Reply:
Did some tests using the center samples for the template but it didn't work too well. It might work but I think we'd also need to play around with the energy window and the cumulative mean normalization.

Reply:
> It might work but I think we'd also need to play around with the energy window and the cumulative mean normalization.

There are some pretty delicate things to get right to make this work, but I think it's worth getting right.  Alternately, we can punt this issue to the future (since this PR already has a lot going on), and just note that boundary effects can be expected when using centered analysis.